File: recursion/towerOfHanoi.cpp
#include <bits/stdc++.h>
// tower of Hanoi using recursion
using namespace std;

void towerOfHanoi(int n, char from, char to, char aux, int &steps)
{
    steps++;
    if (n == 1)
    {
        // cout << "Move plate " << n << " from " << from << " to " << to << " via " << aux << endl;
        return;
    }
    towerOfHanoi(n - 1, from, aux, to, steps);
    // cout << "Move plate " << n << " from " << from << " to " << aux << " via " << aux << endl;
    towerOfHanoi(n - 1, aux, to, from, steps);
    // cout << "Move plate " << n << " from " << aux << " to " << to << " via " << from << endl;
    return;
}

int main()
{
    int n, steps = 0;
    cin >> n;
    towerOfHanoi(n, 'A', 'B', 'C', steps);
    cout << steps;
    return 0;
}

File: Algorithms/Recursion/TowerOfHanoi.java
package Algorithms.Recursion;

public class TowerOfHanoi {
    static int steps = 0;
    public static void helper(int n, int A, int C, int B){
        if(n > 0){
            helper(n - 1, A, B, C);
            System.out.println("move disk " + n + " from rod "+ A +" to rod " + C);
            steps++;
            helper(n - 1, B, C, A);
        }
    }

    public int towerOfHanoi(int n, int A, int C, int B) {
        helper(n, A, C, B);

        return steps;
    }
    public static void main(String[] args) {
        TowerOfHanoi toh = new TowerOfHanoi();
        System.out.println(toh.towerOfHanoi(2, 1, 3, 2));
    }
}
